File: test/VbNet.LanguageServer.Tests/Services/DefinitionServiceTests.cs
using Microsoft.Extensions.Logging.Abstractions;
using VbNet.LanguageServer.Protocol;
using VbNet.LanguageServer.Services;
using VbNet.LanguageServer.Workspace;
using Xunit;

namespace VbNet.LanguageServer.Tests.Services;

/// <summary>
/// Unit tests for DefinitionService.
/// </summary>
public class DefinitionServiceTests
{
    private readonly WorkspaceManager _workspaceManager;
    private readonly DocumentManager _documentManager;
    private readonly DefinitionService _definitionService;

    public DefinitionServiceTests()
    {
        _workspaceManager = new WorkspaceManager(NullLogger<WorkspaceManager>.Instance);
        _documentManager = new DocumentManager(_workspaceManager, NullLogger<DocumentManager>.Instance);
        _definitionService = new DefinitionService(
            _workspaceManager,
            _documentManager,
            NullLogger<DefinitionService>.Instance);

        _workspaceManager.Initialize();
    }

    [Fact]
    public async Task GetDefinitionAsync_NoDocument_ReturnsEmpty()
    {
        var @params = new DefinitionParams
        {
            TextDocument = new TextDocumentIdentifier { Uri = "file:///nonexistent.vb" },
            Position = new Position { Line = 0, Character = 0 }
        };

        var result = await _definitionService.GetDefinitionAsync(@params, CancellationToken.None);

        Assert.Empty(result);
    }

    [Fact]
    public async Task GetDefinitionAsync_StandaloneDocument_ReturnsEmpty()
    {
        var uri = "file:///c:/test/module.vb";
        var text = "Module Module1\n    Sub Main()\n    End Sub\nEnd Module";

        _documentManager.HandleDidOpen(new DidOpenTextDocumentParams
        {
            TextDocument = new TextDocumentItem
            {
                Uri = uri,
                LanguageId = "vb",
                Version = 1,
                Text = text
            }
        });

        var @params = new DefinitionParams
        {
            TextDocument = new TextDocumentIdentifier { Uri = uri },
            Position = new Position { Line = 1, Character = 8 }
        };

        // Without a workspace, we can't get definition info
        var result = await _definitionService.GetDefinitionAsync(@params, CancellationToken.None);

        Assert.Empty(result);
    }

    [Fact]
    public async Task GetDefinitionAsync_NullParams_ReturnsEmpty()
    {
        var result = await _definitionService.GetDefinitionAsync(null!, CancellationToken.None);

        Assert.Empty(result);
    }

    [Fact]
    public async Task GetDefinitionAsync_NullTextDocument_ReturnsEmpty()
    {
        var @params = new DefinitionParams
        {
            TextDocument = null!,
            Position = new Position { Line = 0, Character = 0 }
        };

        var result = await _definitionService.GetDefinitionAsync(@params, CancellationToken.None);

        Assert.Empty(result);
    }
}

File: test/VbNet.LanguageServer.Tests.Vb/Services/DefinitionServiceTests.vb
Imports System.Threading
Imports System.Threading.Tasks
Imports Microsoft.Extensions.Logging.Abstractions
Imports VbNet.LanguageServer.Protocol
Imports VbNet.LanguageServer.Services
Imports VbNet.LanguageServer.Workspace
Imports Xunit

Namespace VbNet.LanguageServer.Tests.Services

    ''' <summary>
    ''' Unit tests for DefinitionService.
    ''' </summary>
    Public Class DefinitionServiceTests
        Private ReadOnly _workspaceManager As WorkspaceManager
        Private ReadOnly _documentManager As DocumentManager
        Private ReadOnly _definitionService As DefinitionService

        Public Sub New()
            _workspaceManager = New WorkspaceManager(NullLogger(Of WorkspaceManager).Instance)
            _documentManager = New DocumentManager(_workspaceManager, NullLogger(Of DocumentManager).Instance)
            _definitionService = New DefinitionService(
                _workspaceManager,
                _documentManager,
                NullLogger(Of DefinitionService).Instance)

            _workspaceManager.Initialize()
        End Sub

        <Fact>
        Public Async Function GetDefinitionAsync_NoDocument_ReturnsEmpty() As Task
            Dim request = New DefinitionParams With {
                .TextDocument = New TextDocumentIdentifier With {.Uri = "file:///nonexistent.vb"},
                .Position = New Position With {.Line = 0, .Character = 0}
            }

            Dim result = Await _definitionService.GetDefinitionAsync(request, CancellationToken.None)

            Assert.NotNull(result)
            Assert.Empty(result)
        End Function

        <Fact>
        Public Async Function GetDefinitionAsync_StandaloneDocument_ReturnsEmpty() As Task
            Dim uri = "file:///c:/test/module.vb"
            Dim text = "Module Module1" & vbLf & "    Sub Main()" & vbLf & "    End Sub" & vbLf & "End Module"

            _documentManager.HandleDidOpen(New DidOpenTextDocumentParams With {
                .TextDocument = New TextDocumentItem With {
                    .Uri = uri,
                    .LanguageId = "vb",
                    .Version = 1,
                    .Text = text
                }
            })

            Dim request = New DefinitionParams With {
                .TextDocument = New TextDocumentIdentifier With {.Uri = uri},
                .Position = New Position With {.Line = 1, .Character = 8}
            }

            Dim result = Await _definitionService.GetDefinitionAsync(request, CancellationToken.None)

            Assert.Empty(result)
        End Function

        <Fact>
        Public Async Function GetDefinitionAsync_NullParams_ReturnsEmpty() As Task
            Dim result = Await _definitionService.GetDefinitionAsync(Nothing, CancellationToken.None)

            Assert.NotNull(result)
            Assert.Empty(result)
        End Function

        <Fact>
        Public Async Function GetDefinitionAsync_NullTextDocument_ReturnsEmpty() As Task
            Dim request = New DefinitionParams With {
                .TextDocument = Nothing,
                .Position = New Position With {.Line = 0, .Character = 0}
            }

            Dim result = Await _definitionService.GetDefinitionAsync(request, CancellationToken.None)

            Assert.NotNull(result)
            Assert.Empty(result)
        End Function
    End Class

End Namespace
